library/core/src/iter/traits/iterator.rs

```diff
@@ -500,8 +500,8 @@ pub trait Iterator {
     /// In other words, it zips two iterators together, into a single one.
     ///
     /// If either iterator returns [`None`], [`next`] from the zipped iterator
-    /// will return [`None`]. If the first iterator returns [`None`], `zip` will
-    /// short-circuit and `next` will not be called on the second iterator.
+    /// will return [`None`]. If the zipped iterator has no more elements to return then attempts
```

Comment:
> no more elements to return

What would this mean for non-fused iterators? What if both iterators return `None` at some point, but both start giving `Some(..)` again some time later? Are you sayingg that the `Zip` iterator might just stop forever at that point (fusing the iterator), or are you just talking about iterators that we know will give `None` forever because of our specializations?

Reply:
A hypothetical implementation could query `size_hint()` of one or both iterators on every call to `next()` and short-circuit when the upper bound of either is 0. That should be safe and not require any specialization. Whether that could be considered as fusing the iterator would depend on how the non-fused one calculates its size hint.

It's intentionally kept vague to leave more flexibility for optimizations while at the same time telling the users that the underlying iterator's `next()` might not be called. Previously it only said that the 2nd iterator's might not be called, now it may call neither.